### DISCUSSION TITLE
How to solve differing number of rows when fitting LGCP

### DISCUSSION THREAD

**Discussion by lvclarke** [inlabru-org/inlabru#191] [Q&A]
Hi,

I'm very new to inlabru and trying to fit my first LGCP model, but keep getting the same error. It looks like it should be a relatively simple fix from the error message. See code below:

# tidy up the data #
mesh <- spatial_list$mesh5
nests <- spatial_list$nests
transects <- spatial_list$transects
boundary <- spatial_list$boundary

# Set experimental mode
bru_options_set(inla.mode = "experimental")

# Specify truncation distance of 80m 
W <- 80

# Specify half-normal detection function 
hn <- function(distance, lsig) {
  exp(-0.5 * (distance / exp(lsig))^2)
}

# Define a (Matern) covariance function for the SPDE
matern <- inla.spde2.pcmatern(mesh,
                              prior.sigma = c(2, 0.01),
                              prior.range = c(50, 0.01)
)

# Specify components of model
cmp <- ~ mySPDE(main = coordinates, model = matern) +
  lsig(1) + Intercept(1)

# Specify formula
formula <- coordinates + distance ~ mySPDE + log_hn(distance, lsig) + Intercept + log(2)

# Fit model
fit <- lgcp(
  components = cmp,
  nests,
  samplers = transects,
  domain = list(
    coordinates = mesh,
    distance = INLA::inla.mesh.1d(seq(0, 8, length.out = 30))
  ),
  formula = form
)

The error message that I keep getting back is:
Error in data.frame(..., check.names = FALSE) : 
  arguments imply differing number of rows: 424, 0

Any help or guidance on this would be greatly appreciated!
Thanks,
Lucy

**Comment by finnlindgren** [comment]
I don't see anything obviously wrong with the code itself, but the error message indicates some data is missing, so first of all check that your `nests`, `transects`, and `mesh` objects contain what you expect them to (and to be safer, use `data = nests` instead of relying on a specific argument order in `lgcp()`).

What `inlabru` version did you use?  Can you try the latest devel version from github? I updated and added some warnings/error messages earlier this week, that might detect the problem sooner and give a clearer message.
The `traceback()` information would also help me identify what part of the code is having trouble.

Since early this year, the new `inla.mode="compact"` is the default in INLA, so you don't need the older "experimental" mode setting anymore.

**Reply by finnlindgren** [reply]
If you have a recent `sp` package and no `rgdal` package installed, you can also try `sp::set_evolution_status(2L)` to work around some issue with the upcoming `rgdal`/`rgeos` package removal. Evolution status `2L` forces `sp` to use the `sf` package for coordinate system operations, which solves issues on systems that don't have `rgdal` installed.

**Reply by lvclarke** [reply]
Hi Finn, 

Thanks so much for such quick help! So I think that my data looks as I would expect/hope:

 head(nests)
           coordinates X.  Transect Number.of.Nests Perpendicular.Distance
1 (514.0718, 1251.641)  1 Sutuba_42               1                   0.03
2 (514.0718, 1251.641) NA Sutuba_42               1                   0.03
3 (514.0718, 1251.641) NA Sutuba_42               1                   0.03
4 (514.0718, 1251.641) NA Sutuba_42               1                   0.03
5 (514.0718, 1251.641) NA Sutuba_42               1                   0.03
6 (514.0718, 1251.641) NA Sutuba_42               1                   0.03

> head(transects)
An object of class "SpatialLinesDataFrame"
Slot "data":
            Name
1   62_Balana_T1
1.1 62_Balana_T1
1.2 62_Balana_T1
1.3 62_Balana_T1
1.4 62_Balana_T1
1.5 62_Balana_T1

Slot "lines":
[[1]]
An object of class "Lines"
Slot "Lines":
[[1]]
An object of class "Line"
Slot "coords":
          [,1]     [,2]
 [1,] 522.3727 1262.799
 [2,] 522.4079 1262.737
 [3,] 522.4447 1262.680
 [4,] 522.4735 1262.598
 [5,] 522.5071 1262.540
 [6,] 522.5983 1262.413
 [7,] 522.6894 1262.269
 [8,] 522.7806 1262.112
 [9,] 522.8510 1261.942
[10,] 522.8415 1261.889



Slot "ID":
[1] "1"


[[2]]
An object of class "Lines"
Slot "Lines":
[[1]]
An object of class "Line"
Slot "coords":
          [,1]     [,2]
 [1,] 522.3727 1262.799
 [2,] 522.4079 1262.737
 [3,] 522.4447 1262.680
 [4,] 522.4735 1262.598
 [5,] 522.5071 1262.540
 [6,] 522.5983 1262.413
 [7,] 522.6894 1262.269
 [8,] 522.7806 1262.112
 [9,] 522.8510 1261.942
[10,] 522.8415 1261.889



Slot "ID":
[1] "1.1"


[[3]]
An object of class "Lines"
Slot "Lines":
[[1]]
An object of class "Line"
Slot "coords":
          [,1]     [,2]
 [1,] 522.3727 1262.799
 [2,] 522.4079 1262.737
 [3,] 522.4447 1262.680
 [4,] 522.4735 1262.598
 [5,] 522.5071 1262.540
 [6,] 522.5983 1262.413
 [7,] 522.6894 1262.269
 [8,] 522.7806 1262.112
 [9,] 522.8510 1261.942
[10,] 522.8415 1261.889



Slot "ID":
[1] "1.2"


[[4]]
An object of class "Lines"
Slot "Lines":
[[1]]
An object of class "Line"
Slot "coords":
          [,1]     [,2]
 [1,] 522.3727 1262.799
 [2,] 522.4079 1262.737
 [3,] 522.4447 1262.680
 [4,] 522.4735 1262.598
 [5,] 522.5071 1262.540
 [6,] 522.5983 1262.413
 [7,] 522.6894 1262.269
 [8,] 522.7806 1262.112
 [9,] 522.8510 1261.942
[10,] 522.8415 1261.889



Slot "ID":
[1] "1.3"


[[5]]
An object of class "Lines"
Slot "Lines":
[[1]]
An object of class "Line"
Slot "coords":
          [,1]     [,2]
 [1,] 522.3727 1262.799
 [2,] 522.4079 1262.737
 [3,] 522.4447 1262.680
 [4,] 522.4735 1262.598
 [5,] 522.5071 1262.540
 [6,] 522.5983 1262.413
 [7,] 522.6894 1262.269
 [8,] 522.7806 1262.112
 [9,] 522.8510 1261.942
[10,] 522.8415 1261.889



Slot "ID":
[1] "1.4"


[[6]]
An object of class "Lines"
Slot "Lines":
[[1]]
An object of class "Line"
Slot "coords":
          [,1]     [,2]
 [1,] 522.3727 1262.799
 [2,] 522.4079 1262.737
 [3,] 522.4447 1262.680
 [4,] 522.4735 1262.598
 [5,] 522.5071 1262.540
 [6,] 522.5983 1262.413
 [7,] 522.6894 1262.269
 [8,] 522.7806 1262.112
 [9,] 522.8510 1261.942
[10,] 522.8415 1261.889



Slot "ID":
[1] "1.5"



Slot "bbox":
        min      max
x  522.3727  522.851
y 1261.8889 1262.799

Slot "proj4string":
Coordinate Reference System:
Deprecated Proj.4 representation:
 +proj=utm +zone=28 +datum=WGS84 +units=km +no_defs 
WKT2 2019 representation:
PROJCRS["unknown",
    BASEGEOGCRS["unknown",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]],
            ID["EPSG",6326]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8901]]],
    CONVERSION["UTM zone 28N",
        METHOD["Transverse Mercator",
            ID["EPSG",9807]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",-15,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9996,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",500000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]],
        ID["EPSG",16028]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["kilometre",1000,
                ID["EPSG",9036]]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["kilometre",1000,
                ID["EPSG",9036]]]] 

The inlabru version I hade been using is: 
packageDescription("inlabru")$Version
[1] "2.7.0"

I then downloaded development version from github as suggested, and re ran the code.

I now get the following error message when trying to fit the LGCP:
Error in as.data.frame.default(x[[i]], optional = TRUE) : 
  cannot coerce class ‘structure("standardGeneric", package = "methods")’ to a data.frame

Seems that there must be an issue with data inputs at some stage? Also tried sp::set_evolution_status(2L), but get the same error message. 

Using traceback():
traceback()
12: stop(gettextf("cannot coerce class %s to a data.frame", sQuote(deparse(class(x))[1L])), 
        domain = NA)
11: as.data.frame.default(x[[i]], optional = TRUE)
10: as.data.frame(x[[i]], optional = TRUE)
9: (function (..., row.names = NULL, check.rows = FALSE, check.names = TRUE, 
       fix.empty.names = TRUE, stringsAsFactors = FALSE) 
   {
       data.row.names <- if (check.rows && is.null(row.names)) 
           function(current, new, i) {
               if (is.character(current)) 
                   new <- as.character(new)
               if (is.character(new)) 
                   current <- as.character(current)
               if (anyDuplicated(new)) 
                   return(current)
               if (is.null(current)) 
                   return(new)
               if (all(current == new) || all(current == "")) 
                   return(new)
               stop(gettextf("mismatch of row names in arguments of 'data.frame', item %d", 
                   i), domain = NA)
           }
       else function(current, new, i) {
           if (is.null(current)) {
    ...
8: do.call(data.frame, c(x, alis))
7: as.data.frame.list(response[!idx])
6: as.data.frame(response[!idx])
5: data.frame(..., check.names = FALSE)
4: cbind(deparse.level, ...)
3: cbind(data, as.data.frame(response[!idx]))
2: like(family = "cp", formula = formula, data = data, samplers = samplers, 
       E = E, ips = ips, domain = domain, ..., options = options)
1: lgcp(components = cmp, data = nests, samplers = transects, domain = list(coordinates = mesh, 
       distance = INLA::inla.mesh.1d(seq(0, 8, length.out = 30))), 
       formula = formula)
> 
Are you able to identify from this where the issue might be?

Many thanks again for all of your help! Much appreciated!
Lucy

**Reply by finnlindgren** [reply]
The error message/traceback does indicate a potential problem with the code in some situations to do with unnecessary row names, but in this particular case I think that’s just a follow-on error.
Your nests variable has a Perpendicular.Distance column, but in your model definition code it’s assumed that the variable is called “distance”. Try `nests$distance <- nests$Perpendicular.Distance` and rerun the model.

**Reply by lvclarke** [reply]
Hi Finn - apologies that is a silly error with the data on my part.

Rerunning the code I now get the following error message:

 *** inla.core.safe:  inla.program has crashed: rerun to get better initial values. try=2/2 
Assertion failed: (0 == 1), function GMRFLib_idxval_nsort_x_core, file idxval.c, line 846.
sh: line 1: 57561 Abort trap: 6           '/Library/Frameworks/R.framework/Versions/4.2/Resources/library/INLA/bin/mac/64bit/inla.run' -s -v -t2:1 -B0 -P compact '/private/var/folders/k6/2_1lhr252935zkghmn403sw00000gn/T/Rtmp7fJ2H0/filec967432b2017/Model.ini' > '/private/var/folders/k6/2_1lhr252935zkghmn403sw00000gn/T/Rtmp7fJ2H0/filec967432b2017/Logfile.txt'
Error in inla.inlaprogram.has.crashed() : 
  The inla-program exited with an error. Unless you interupted it yourself, please rerun with verbose=TRUE and check the output carefully.
  If this does not help, please contact the developers at <help@r-inla.org>.
Error: C stack usage  8649917 is too close to the limit

Have you seen this issue before? 

Thanks again,
Lucy

**Reply by finnlindgren** [reply]
Looks like you’re running on a Mac? What INLA version? The most recent inla builds are for R4.3, so you might need to either use an older build for R4.2, or upgrade R to 4.3 (and if you do that you need to upgrade/reinstall all ar packages).

**Reply by lvclarke** [reply]
Hi Finn, 

Yes - running on rather old mac I'm afraid... I can only get version R 4.2.3 at the moment without doing a full MacOS software update. I've gone back down to inlabru version 2.5.0. When I now run the code I get the following:

*** max_correction = 25.01 >= 25.00, so 'vb.correction' is aborted
*** Please (re-)consider your model, priors, confounding, etc.
*** You can change the emergency value (current value=25.00) by 
*** 	'control.inla=list(control.vb=list(emergency=...))'

 the code continues to run with lines looking like below, and then the same error message pops up:

GMRFLib_2order_approx: rescue NAN/INF values in logl for idx=12540
GMRFLib_2order_approx: rescue NAN/INF values in logl for idx=12541
GMRFLib_2order_approx: rescue NAN/INF values in logl for idx=12542
GMRFLib_2order_approx: rescue NAN/INF values in logl for idx=12543

Assertion failed: (arg->Q->a[0] >= 0.0), function GMRFLib_tabulate_Qfunc_core, file tabulate-Qfunc.c, line 147.
sh: line 1: 72251 Abort trap: 6           '/Library/Frameworks/R.framework/Versions/4.2/Resources/library/INLA/bin/mac/64bit/inla.run' -s -v -t2:1 -B0 -P compact '/private/var/folders/k6/2_1lhr252935zkghmn403sw00000gn/T/RtmplTMjEt/fileea6037736d99/Model.ini' > '/private/var/folders/k6/2_1lhr252935zkghmn403sw00000gn/T/RtmplTMjEt/fileea6037736d99/Logfile.txt'
Error in inla.inlaprogram.has.crashed() : 
  The inla-program exited with an error. Unless you interupted it yourself, please rerun with verbose=TRUE and check the output carefully.
  If this does not help, please contact the developers at <help@r-inla.org>.
Error: C stack usage  8890166 is too close to the limit

Do I need to go to a lower version of inlabru?

**Reply by lvclarke** [reply]
Sorry for so many questions - really appreciate your help!

**Reply by finnlindgren** [reply]
The inlabru version should still be the latest one; older ones won’t help! It’s the INLA package version that may need to be a couple of weeks old.

**Reply by lvclarke** [reply]
Ok sorry my mistake! I've reverted back to version 2.7.0 of inlabru, and have installed an older version of INLA but now seem to be getting this error:

    packageDescription("INLA")$Version
    [1] "22.12.16"

    fit <- lgcp(
    components = cmp,
    data = nests,
    samplers = transects,
    domain = list(
    coordinates = mesh,
    distance = INLA::inla.mesh.1d(seq(0, 8, length.out = 30))
    ),
    formula = form,
    )
   
Error in inla.call.builtin() : INLA installation error; no such file

Is the INLA package incorrectly installed?

**Reply by lvclarke** [reply]
I've also tried to revert back as all my code had stopped working using:
install.packages("INLA",repos=c(getOption("repos"),INLA="https://inla.r-inla-download.org/R/stable"), dep=TRUE) 
but also having no luck with this - not sure what is now going on! 

Warning in install.packages :
  dependencies ‘Rgraphviz’, ‘graph’ are not available
trying URL 'https://inla.r-inla-download.org/R/stable/bin/macosx/contrib/4.2/INLA_23.04.24.tgz'
Content type 'application/x-tar' length 180725988 bytes (172.4 MB)
==
downloaded 7.5 MB

Warning in install.packages :
  downloaded length 7831866 != reported length 180725988
Warning in install.packages :
  URL 'https://inla.r-inla-download.org/R/stable/bin/macosx/contrib/4.2/INLA_23.04.24.tgz': Timeout of 60 seconds was reached
Error in download.file(url, destfile, method, mode = "wb", ...) : 
  download from 'https://inla.r-inla-download.org/R/stable/bin/macosx/contrib/4.2/INLA_23.04.24.tgz' failed
Warning in install.packages :
  download of package ‘INLA’ failed

**Reply by finnlindgren** [reply]
The warning about Rgraphviz and graph is fine; they are bioconductor packages not on CRAN.

for the inla download you hit the default R timeout of 60 seconds. Increase it with  options(timeout=99999)

**Reply by lvclarke** [reply]
Great thanks - this has worked to revert me back to INLA package 23.04.24. I've re-run all my code but still getting the following error:

Assertion failed: (arg->Q->a[0] >= 0.0), function GMRFLib_tabulate_Qfunc_core, file tabulate-Qfunc.c, line 147.
sh: line 1: 79531 Abort trap: 6           '/Library/Frameworks/R.framework/Versions/4.2/Resources/library/INLA/bin/mac/64bit/inla.run' -s -v -t2:1 -B0 -P compact '/private/var/folders/k6/2_1lhr252935zkghmn403sw00000gn/T/RtmpNFJGQ4/file12c1454bc4dc1/Model.ini' > '/private/var/folders/k6/2_1lhr252935zkghmn403sw00000gn/T/RtmpNFJGQ4/file12c1454bc4dc1/Logfile.txt'
Error: C stack usage  17562429 is too close to the limit

I tried to move to an older verison of INLA but unsure if I did it correctly given the errors I started to get below:
Error in inla.call.builtin() : INLA installation error; no such file

Are you able to advice on how best to proceed? Apologies for all of the questions - at a bit of a loss for what to try!

Thanks,
Lucy

**Reply by finnlindgren** [reply]
Seems you may be running out of memory… what’s the size of the model? Number of mesh nodes (mesh$n) and number of transects?

**Reply by lvclarke** [reply]
Hi Finn, 

The mesh has 3619 nodes and there are 539 transects (64 walked on multiple occasions).

**Reply by finnlindgren** [reply]
That shouldn’t be a problem, and the model structure is quite simple, and comparable to our examples. Would you be able to email me (Finn.Lindgren@gmail.com) the data files and a code script that reads the data and runs your model? That would make it easier for me to debug.

**Reply by lvclarke** [reply]
Thanks very much Finn - will do!
